File: scripts/core/logger.py
import os
import sys
import inspect
from datetime import datetime, timedelta
from pathlib import Path

LOG_DIR = None
_init_done = False
_log_file = None


def _resolve_log_dir():
    global LOG_DIR
    if LOG_DIR is None:
        try:
            from adapter.bridge.paths import get_data_dir
            LOG_DIR = os.path.join(get_data_dir(), "logs")
        except Exception:
            LOG_DIR = os.path.join(os.path.expanduser("~"), ".remora", "logs")
    return LOG_DIR


def init():
    global _init_done, _log_file

    log_dir = _resolve_log_dir()
    os.makedirs(log_dir, exist_ok=True)

    today_str = datetime.now().strftime("%Y-%m-%d")
    log_path = os.path.join(log_dir, "system.log")

    if os.path.exists(log_path):
        mtime = datetime.fromtimestamp(os.path.getmtime(log_path))
        if mtime.strftime("%Y-%m-%d") != today_str:
            archive_path = os.path.join(log_dir, f"system.{mtime.strftime('%Y-%m-%d')}.log")
            os.rename(log_path, archive_path)

    _log_file = log_path
    _init_done = True

    # Cleanup old logs (>3 days)
    cutoff = datetime.now() - timedelta(days=3)
    try:
        for fname in os.listdir(log_dir):
            if fname.startswith("system.") and fname.endswith(".log") and fname != "system.log":
                fpath = os.path.join(log_dir, fname)
                try:
                    ftime = datetime.fromtimestamp(os.path.getmtime(fpath))
                    if ftime < cutoff:
                        os.remove(fpath)
                except Exception:
                    pass
    except Exception:
        pass


def _format_caller():
    frame = inspect.currentframe()
    try:
        caller = frame.f_back.f_back.f_back
        if caller is None:
            caller = frame.f_back.f_back
        filename = os.path.basename(caller.f_code.co_filename)
        lineno = caller.f_lineno
        return f"{filename}:{lineno}"
    finally:
        del frame


def _log(level, msg):
    global _init_done, _log_file
    if not _init_done:
        init()
    if _log_file is None:
        return

    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    caller = _format_caller()
    line = f"[{timestamp}] [{level}] [{caller}] {msg}\n"

    try:
        with open(_log_file, "a", encoding="utf-8") as f:
            f.write(line)
    except Exception:
        pass


def _write_raw(log_path, content, max_bytes=1024*1024):
    try:
        path = Path(log_path)
        if path.exists() and path.stat().st_size > max_bytes:
            with open(path, "w", encoding="utf-8") as f:
                f.write(f"=== Log Rotated at {datetime.now().isoformat()} ===\n")
        with open(path, "a", encoding="utf-8") as f:
            f.write(content)
    except Exception:
        pass


def info(msg):
    _log("INFO", msg)


def warn(msg):
    _log("WARN", msg)
    print(f"[WARN] {msg}", file=sys.stderr)


def error(msg):
    _log("ERROR", msg)
    print(f"[ERROR] {msg}", file=sys.stderr)


def profile(msg, log_path=None):
    if log_path is not None:
        _write_raw(log_path, msg)
    elif isinstance(msg, str) and msg.strip().startswith("==="):
        try:
            from adapter.bridge.paths import HOOKS_PROFILE_LOG
            _write_raw(HOOKS_PROFILE_LOG, msg)
        except Exception:
            pass
    else:
        _log("PROF", msg)

File: packages/core/src/logger.ts
/**
 * Structured file logger with trace ID, rotation, and profiling.
 *
 * Python equivalent: scripts/core/logger.py
 * Platform-agnostic — uses node:fs, no adapter dependencies.
 */

import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { randomUUID } from "node:crypto";

// ── Configuration ────────────────────────────────────────────────

const LOG_DIR = process.env.REMORA_LOG_DIR ?? path.join(os.tmpdir(), "remora", "log");
const MAX_AGE_DAYS = 3;
let traceId = process.env.REMORA_TRACE_ID ?? `s_${randomUUID().slice(0, 8)}`;

const HOOKS_PROFILE_LOG =
  process.env.REMORA_HOOKS_PROFILE_LOG ??
  path.join(os.homedir(), ".remora", "data", "hooks_profile.log");

const LEVEL_ENV = (process.env.REMORA_LOG_LEVEL ?? "INFO").toUpperCase();
const LEVELS: Record<string, number> = { DEBUG: 0, INFO: 1, WARN: 2, ERROR: 3, OFF: 4 };
const level = LEVELS[LEVEL_ENV] ?? 1;

let initDone = false;
let logFile: string | null = null;

// ── Public API ───────────────────────────────────────────────────

export function setTraceId(tid: string): void {
  traceId = tid;
  process.env.REMORA_TRACE_ID = tid;
}

export function init(): void {
  if (initDone) return;

  const envTid = process.env.REMORA_TRACE_ID;
  if (envTid) setTraceId(envTid);

  fs.mkdirSync(LOG_DIR, { recursive: true });

  const now = new Date();
  const todayStr = now.toISOString().slice(0, 10);
  const logPath = path.join(LOG_DIR, "system.log");

  if (fs.existsSync(logPath)) {
    const mtime = fs.statSync(logPath).mtime;
    const mtimeDay = mtime.toISOString().slice(0, 10);
    if (mtimeDay !== todayStr) {
      const archivePath = path.join(LOG_DIR, `system.${mtimeDay}.log`);
      fs.renameSync(logPath, archivePath);
    }
  }

  logFile = logPath;
  initDone = true;

  // Cleanup old archived logs
  const cutoff = Date.now() - MAX_AGE_DAYS * 86400_000;
  try {
    for (const fname of fs.readdirSync(LOG_DIR)) {
      if (fname.startsWith("system.") && fname.endsWith(".log") && fname !== "system.log") {
        const fpath = path.join(LOG_DIR, fname);
        try {
          if (fs.statSync(fpath).mtimeMs < cutoff) {
            fs.unlinkSync(fpath);
          }
        } catch {
          // ignore individual file errors
        }
      }
    }
  } catch {
    // ignore directory read errors
  }
}

export function debug(msg: string): void {
  _log("DEBUG", msg);
}

export function info(msg: string): void {
  _log("INFO", msg);
}

export function warn(msg: string): void {
  _log("WARN", msg);
  process.stderr.write(`[WARN] ${msg}\n`);
}

export function error(msg: string): void {
  _log("ERROR", msg);
  process.stderr.write(`[ERROR] ${msg}\n`);
}

export function profile(msg: string, logPath?: string | null): void {
  if (logPath != null) {
    _writeRaw(logPath, msg);
  } else if (msg.trimStart().startsWith("===")) {
    try {
      _writeRaw(HOOKS_PROFILE_LOG, msg);
    } catch {
      // ignore
    }
  } else {
    _log("PROF", msg);
  }
}

// ── Internal ─────────────────────────────────────────────────────

function _shouldLog(lvl: string): boolean {
  return (LEVELS[lvl] ?? 1) >= level;
}

function _formatCaller(): string {
  const trace = new Error().stack;
  if (!trace) return "unknown:0";
  // stack format: "Error\n    at func (file:line:col)" or "    at file:line:col"
  // We want frame 5: _formatCaller → _log → info/warn/... → caller
  const lines = trace.split("\n");
  const callerLine = lines[5] ?? lines[lines.length - 1] ?? "";
  const match = callerLine.match(/(?:\()?([^:)]+):(\d+)(?::\d+)?\)?$/);
  if (match) {
    return `${path.basename(match[1])}:${match[2]}`;
  }
  return "unknown:0";
}

function _log(lvl: string, msg: string): void {
  if (!_shouldLog(lvl)) return;
  if (!initDone) init();
  if (!logFile) return;

  const ts = new Date().toISOString().replace("T", " ").slice(0, 19);
  const caller = _formatCaller();
  const line = `[TID:${traceId}] [${ts}] [${lvl.padEnd(5)}] [${caller}] ${msg}\n`;

  try {
    fs.appendFileSync(logFile, line, "utf-8");
  } catch {
    // ignore write errors
  }
}

function _writeRaw(logPath: string, content: string, maxBytes: number = 1024 * 1024): void {
  try {
    if (fs.existsSync(logPath)) {
      const stat = fs.statSync(logPath);
      if (stat.size > maxBytes) {
        fs.writeFileSync(logPath, `=== Log Rotated at ${new Date().toISOString()} ===\n`, "utf-8");
      }
    }
    fs.appendFileSync(logPath, content, "utf-8");
  } catch {
    // ignore
  }
}
